acorn/src/expression.js

```diff
@@ -260,29 +260,38 @@ pp.parseExprSubscripts = function(refDestructuringErrors) {
 pp.parseSubscripts = function(base, startPos, startLoc, noCalls) {
   let maybeAsyncArrow = this.options.ecmaVersion >= 8 && base.type === "Identifier" && base.name === "async" &&
       this.lastTokEnd === base.end && !this.canInsertSemicolon() && this.input.slice(base.start, base.end) === "async"
+  let shortCircuited = false
   while (true) {
-    let element = this.parseSubscript(base, startPos, startLoc, noCalls, maybeAsyncArrow)
+    let element = this.parseSubscript(base, startPos, startLoc, noCalls, maybeAsyncArrow, shortCircuited)
+    if (this.options.ecmaVersion >= 11 && element.optional) shortCircuited = true
     if (element === base || element.type === "ArrowFunctionExpression") return element
     base = element
   }
 }
 
-pp.parseSubscript = function(base, startPos, startLoc, noCalls, maybeAsyncArrow) {
```

Comment:
@mysticatea Not sure if it breaks plugins, but you can get rid of the extra func param here if you move  `parseSubscript`  into the while loop. And rename `parseSubscripts`  to `parseSubscript`  to avoid possible plugin conflict. That way you could use a `var isOptional` at the top before the iteration loop. You can get the ideas from [v8](https://github.com/v8/v8/blob/master/src/parsing/parser-base.h#L3168)

Reply:
I think it's a breaking change for plugins because plugins can override Parser methods.

Reply:
Agreed, let's try not to break the plugin interface!